enhancements/machine-api/defaulting-validating.md

```diff
@@ -0,0 +1,137 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+Today the machine API embeds multiple particular cloud provider APIs as a single raw extension within the machine resource.
+
+This results in a lack of first class support for manipulating these APIs from a Kubernetes point of view, particularly in a lack of defaulting and validating capabilities.
+
+This proposal outlines a mechanism to enable defaulting and validation on the particular providers API before the input is persisted to etcd.
+
+## Motivation
+
+Having multiple APIs as a raw extension is brittle and let as swallow any bad user input and only fail in the backend.
+Failing earlier and communicating it meaningfully is better. In this line we want to ensure that API defaulting and validation happens before a resource input is persisted to disk to dramatically improve the UX for manipulating resources under the `machine.openshift.io`  group.
+
+### Goals
+
+- Enable defaulting and validating on the particular providerSpec for the Machine resource.
+
+### Non-Goals
+
+- Introduce API breaking changes.
+- Externalise the providerSpec and expose a new CRD.
+- Flesh out the fields and details for each particular provider defaults and validation. That can be carefully discussed in each individual PR and updated back to this proposal in retrospective.
+
+## Proposal
+
+Today the machine API embeds the particular cloud provider API as a raw extension.
+
+```
+// ProviderSpec defines the configuration to use during node creation.
+type ProviderSpec struct {
+  Value *runtime.RawExtension `json:"value,omitempty"`
+}
+```
+
+This proposes to run both a `MutatingWebhookConfiguration` and a `ValidatingWebhookConfiguration` over the machine resource to let each provider define their API defaulting and validation.
+
+### User Stories [optional]
+
+#### Story 1
+
+As an operator I want the product to communicate meaningfully when my input is bad before it gets persisted so I can fix it and save time.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Plumbing
+- The Machine API Operator (MAO) will expose both a `MutatingWebhookConfiguration` and a `ValidatingWebhookConfiguration` to the Cluster Version Operator (CVO).
+
+- The MAO will expose to the CVO the a `Service` referenced in the webhook config that will resolve to the `Endpoint` where the webhook is running.
+
+- The webhook server can be managed by the same manager that manages the machineSet controller. This is a convenient place so it will let us manage single `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` resources that can be extended with multiple `paths` for other machine API resources, e.g `machineSet`, `machineHealthCheck`, etc. All being reachable by the same `Service` resolving to the same `Endpoint`.
```

Comment:
Upstream run their webhooks as separate pods, I'm not entirely sure why. But perhaps we should run them as a separate container within mapi-controllers, this would make reviewing logs easier as the webhook logs would be separate from the controller logs, not sure if there are other advantages to that?

Reply:
cool, let's explore it see if we identify more advantages.

Reply:
Based on https://kubernetes.slack.com/archives/C8TSNPY4T/p1591171775388700
Discussed with Joel and decided to go with it on the  existing manager for now.